src/model/dfcxx/lib/dfcxx/graph.cpp

```diff
@@ -62,4 +69,86 @@ void Graph::addChannel(DFVariableImpl *source, DFVariableImpl *target,
   }
 }
 
+void Graph::transferFrom(Graph &&graph) {
+  nodes.merge(std::move(graph.nodes));
+  inputs.merge(std::move(graph.inputs));
+  outputs.merge(std::move(graph.outputs));
+  connections.merge(std::move(graph.connections));
+}
+
+Node Graph::getNodeByName(const std::string &name) {
+  auto it = nameMap.find(name);
+  if (it == nameMap.end()) {
+    throw new std::invalid_argument("Non-existent node with name: " + name);
+  }
+  return it->second;
+}
+
+void Graph::resetNodeName(const std::string &name) {
+  auto it = nameMap.find(name);
+  if (it == nameMap.end()) {
+    throw new std::invalid_argument("Non-existent node with name: " + name);
+  }
+  DFVariableImpl *ptr = it->second.var;
+  nameMap.erase(it);
+  ptr->resetName();
+}
+
+void Graph::deleteNode(Node node) {
+  connections.erase(node);
+  inputs.erase(node);
+  outputs.erase(node);
+  startNodes.erase(node);
+  nodes.erase(node);
+}
+
+void Graph::rebindInput(Node source, Node input, Graph &graph) {
+  auto &conns = graph.connections;
+  for (auto &out: graph.outputs[input]) {
+    for (auto &in: graph.inputs[out.target]) {
+      if (in.source == input && out == in) {
+        in.source = source;
+        outputs[source].push_back(in);
+        break;
+      }
+    }
+    auto it = conns.find(out.target);
+    if (it != conns.end() && it->second.source == input) {
+      it->second.source = source;
+    }
+  }
+}
+
+Node Graph::rebindOutput(Node output, Node target, Graph &graph) {
+  auto &inSrc = graph.inputs[output].front().source;
+  auto &outs = graph.outputs[inSrc];
+  for (auto it = outs.begin(); it != outs.end(); ++it) {
+    if (it->target == output) {
+      if (target.type == OpType::NONE) {
+        outs.erase(it);
+        for (auto &out: outputs[target]) {
+          for (auto &in: inputs[out.target]) {
+            if (in.source == target && out == in) {
+              in.source = it->source;
+              outs.push_back(in);
+            }
+          }
+          auto conIt = connections.find(out.target);
+          if (conIt != connections.end() && conIt->second.source == target) {
+            conIt->second.source = it->source;
+          }
+        }
+        target = it->source;
+      } else {
+        it->target = target;
+        inputs[target].clear();
+        inputs[target].push_back(*it);
+        connections[target] = *it;
+      }
+      break;
+    }

```

Comment:
This code fragment is too "nesting". Please decrease it, using "continue" operator inside loop, etc.